chain/chunks/src/lib.rs

```diff
@@ -1202,79 +1226,163 @@ impl ShardsManager {
+    ///    decode and persist the full chunk
+    ///
+    /// Params
+    /// `partial_encoded_chunk`: the partial encoded chunk needs to be processed. `MaybeValidated`
+    ///                          denotes whether the chunk header has been validated or not
+    /// `chain_head`: current chain head, used for validating chunk header
+    ///
+    /// Returns
+    ///  ProcessPartialEncodedChunkResult::Known: if all information in
+    ///    the `partial_encoded_chunk` is already known and no further processing is needed
+    ///  ProcessPartialEncodedChunkResult::NeedBlock: if the previous block is needed
+    ///    to finish processing
+    ///  ProcessPartialEncodedChunkResult::NeedMorePartsOrReceipts: if more parts and receipts
+    ///    are needed for processing the full chunk
+    ///  ProcessPartialEncodedChunkResult::HaveAllPartsAndReceipts: if all parts and
+    ///    receipts in the chunk are received and the chunk has been processed.
+    pub fn process_partial_encoded_chunk(
+        &mut self,
+        partial_encoded_chunk: MaybeValidated<&PartialEncodedChunkV2>,
+        chain_head: Option<&Tip>,
+        chain_store: &mut ChainStore,
+        rs: &mut ReedSolomonWrapper,
+    ) -> Result<ProcessPartialEncodedChunkResult, Error> {
         let header = &partial_encoded_chunk.header;
         let chunk_hash = header.chunk_hash();
-
-        // 2. Leave if we received known chunk
+        // Verify the partial encoded chunk is valid and worth processing
+        // 1.a Leave if we received known chunk
+        // Note that we have to still process the chunk even if all parts and receipts included
+        // in `partial_encoded_chunk` are already in `encoded_chunks`. This is because
+        // `try_process_full_chunk` can only succ
         if let Some(entry) = self.encoded_chunks.get(&chunk_hash) {
-            let know_all_parts = partial_encoded_chunk
-                .parts
-                .iter()
-                .all(|part_entry| entry.parts.contains_key(&part_entry.part_ord));
-
-            if know_all_parts {
-                let know_all_receipts = partial_encoded_chunk
-                    .receipts
-                    .iter()
-                    .all(|receipt| entry.receipts.contains_key(&receipt.1.to_shard_id));
-
-                if know_all_receipts {
-                    return Ok(ProcessPartialEncodedChunkResult::Known);
-                }
+            if entry.complete {
+                return Ok(ProcessPartialEncodedChunkResult::Known);
             }
         };
 
-        // 3. Checking chunk height
+        // 1.b Checking chunk height
         let chunk_requested = self.requested_partial_encoded_chunks.contains_key(&chunk_hash);
         if !chunk_requested {
             if !self.encoded_chunks.height_within_horizon(header.height_created()) {
-                return Err(Error::ChainError(ErrorKind::InvalidChunkHeight.into()));
+                return Ok(ProcessPartialEncodedChunkResult::Known);
```

Comment:
why? Does this change allow for an exploit given that we are not doing a protocol change here?

Reply:
I don't think we ever handle the error returned by this function anyways. 

I can make a separate PR for this. When I made the PR, I didn't know the issue will occur on mainnet again so I included multiple changes related to processing partial encoded chunks

Reply:
I'd add a comment to mention why we return "Known" (as I'd have expected InvalidChunkHeight as it was before)

Reply:
Yeah that would make more sense